### PR TITLE
feat(lookup): add lookup

### PR DESCRIPTION
**Pull request by MaximBalaganskiy**
Demo is at the bottom
https://maximbalaganskiy.github.io/aurelia-ux-demo/#/components

The dropdown pops up at the top if there is little space below.

### CODE REVIEW

**Comment by ben-girardet** [comment]
Love it. Great work.

I've done a quick tryout on your demo. I think it would help to have a little margin with the window. That would increase readability of the popup. Would be amazing to have it as a themable property

Capture d’écran 2020-05-20 à 10 12 55

Also a question, what is the strategy for lookup with hundreds or thousands of options ? Would be helpful to have a little help regarding the usage that you have put in place. Seems like you have make it configurable but didn't understand yet.

**Comment by MaximBalaganskiy** [comment]
There is a templatable error part. If there are too many records a consumer would throw an error in the options function and this would display an error option with an error message

**Comment by MaximBalaganskiy** [comment]
The component supports either an array or a function returning a promise for an array

**Comment by MaximBalaganskiy** [comment]
@ben-girardet added more theme properties

One thing lookup and select still suffer from is scrolling - wheel event is fine, but not scrolling with the bar. `window.scroll` does not happen because `router-view` is being scrolled.

**Comment by MaximBalaganskiy** [comment]
Angular solves it by adding a backdrop which covers the whole window and catches all scrolling events.

**Comment by MaximBalaganskiy** [comment]
Haha, easy solution - capture the event :)

**Comment by bigopon** [comment]
@MaximBalaganskiy @ben-girardet i think for the list positioning, we better use the positioning service that was recently added. Thoughts?

Also, we need to take care of resize/orientation change event as well, to re-position the list

File: packages/lookup/src/ux-lookup.ts [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+        top: undefined,
+        bottom: `${window.innerHeight - availableHeight - windowEdgeDistance}px`,
+        maxHeight: availableHeight,
+        width: inputRect.width
+      };
+    }
+  }
+
+  handleEvent(evt: Event): void {
+    if (evt.currentTarget === this.inputElement) {
```
Comment (bigopon) [inline packages/lookup/src/ux-lookup.ts:188]: it's a weird looking combination of if and switch, though I like how it's organised

Comment (MaximBalaganskiy) [reply packages/lookup/src/ux-lookup.ts]: fixed

**Comment by MaximBalaganskiy** [comment]
@MaximBalaganskiy not sure which positioning service you have in mind

**Comment by MaximBalaganskiy** [comment]
@bigopon anchor is now updated on resize. Mind you, in some cases this causes scrolling which hides the popup

**Comment by bigopon** [comment]
Here is the PR that added it #254 

For scrolling that hides popup, we can follow native on this I think. At the moment, any moment you scroll outside (we can capture this event) the popup of a `<select/>` hides immediately.

**Comment by MaximBalaganskiy** [comment]
Then scrolling is already like native

**Comment by bigopon** [comment]
Oh I see what you meant. We can do extra bit (not sure 100%): after resizing/moving, we call `optionElement.scrollIntoView()`?

**Comment by MaximBalaganskiy** [comment]
Would need trickery to suppress scroll event

**Comment by bigopon** [comment]
> Mind you, in some cases this causes scrolling which hides the popup

Did you meant when moving the popup/resizing it, sometimes it cause the popup to scroll and hide currently selected option?

**Comment by MaximBalaganskiy** [comment]
When window is resized the content is scrolled and popup closes

File: packages/lookup/src/discardable-promise.ts [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    this.isDiscarded = true;
+  }
+}
+
+/**
+ * Sets the internal state of a promise to discarded
+ * @param discardable A promise to discard
+ */
+export function discard<T>(discardable: DiscardablePromise<T> | null | undefined) {
+  if (discardable) {
```
Comment (bigopon) [inline packages/lookup/src/discardable-promise.ts:32]: we are on TS v3.8.3, can just dop `discardable?.discard()`

File: packages/lookup/src/ux-lookup.ts [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+    }
+    this.updateAnchor();
+    windowEvents.map(x => window.addEventListener(x, this, true));
+    this.taskQueue.queueTask(() => this.isOpen = true);
+  }
+
+  close() {
+    this.isOpen = false;
+    this.focusedOption = undefined;
+    windowEvents.map(x => window.addEventListener(x, this, true));
```
[the remark this answers is not in the document]
Comment (MaximBalaganskiy) [reply packages/lookup/src/ux-lookup.ts]: done

File: packages/lookup/src/ux-lookup.ts [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+  bind() {
+    this.themeChanged(this.theme);
+    this.valueFieldChanged();
+    this.displayFieldChanged();
+    this.optionsChanged();
+  }
+
+  attached() {
+    this.inputElement = this.element.parentElement?.querySelector<UxInputElement>('ux-input');
```
[the remark this answers is not in the document]
Comment (MaximBalaganskiy) [reply packages/lookup/src/ux-lookup.ts]: added input

**Comment by bigopon** [comment]
Thanks @MaximBalaganskiy for the nice work

**Pull request merged**